### PR TITLE
grSim Implementing Kick and Chip Primitives

### LINKED ISSUE #93: grSim backend - Implement Kick and Chip Primitives
[grSim](https://github.com/RoboCup-SSL/grSim) is the "standard" simulator for the SSL Division, and is the simulator we use.

Because we currently don't/can't simulate our robot's firmware, we need to re-create how we move and control the robots for the simulator. Rather than sending a Primitive command to a robot over radio, which will then be executed by the robot firmware in order to do whatever the command asks, we need to send different commands to the simulator to move the robots.

We already have the ability to send commands to the simulator with #12 . The ability to control a robot's movement is further abstracted by the motion controller developed in #9 . This imitates the bang-bang motion controller on our robots, so that robot's will speed up and slow down more realistically as the move in the simulator.

**Currently blocked by #9 **

The task here is to create a function/routine to implement our Shoot Primitive using this motion-controller abstraction, so this Primitive can be simulated. We need these primitive to work in the simulator so we can actually use the simulator for testing.

The spec for the Shoot Primitive can be found in #86 , and also in the ShootPrimitive class from #87 

The ShootPrimitive class will be the input. The exact place this code will go will be revised once #9 is merged, and we will update the ticket.

### PR DESCRIPTION
**Pull request by jessicahe21**
### Description

Implemented kick and chip primitives for grSim.

1. If the robot is in line with the shot (less than two robot radius within the line in the direction of the shot), can go for the shot at the given kick/chip origin
2. If not, the robot will move to the closest point on the line from its current position, then check for 1) again. (Sort of looks like a 90 degree turn)

This is a simplified implementation of the originally described characteristics in #86 , which was more of a parabola pathing that consisted of many small line segments. 

- Added `closestPointOnLine` function under `geom/util` that takes a line as input (the existing one takes segments only).

- Modified `offSetToLine` so that its return value, distance to line, is always positive.

### Testing Done
In grSim, the robot is able to reproduce steps 1) and 2) from various different angles as well as different kick/chip origins.

### Resolved Issues
Resolves #93 

### Review Checklist
- [x] **Start of document comments**: each `.cpp` and `.h` file should have a comment at the start of it. See files in the `thunderbots/software/geom` folder for examples.
- [x] **Function comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the classes defined in `thunderbots/software/geom` 
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/thunderbots/software/grsim_communication/main.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/grsim_communication/main.cpp]: I'm not sure if this will compile? You'll probably want to provide some default arguments here, like:
`Ball ball = Ball(Point(), Vector());`

**Review by MathewMacDougall — COMMENTED** [review]
Ah I just realized you actually don't need to know about the ball's state :see_no_evil:. I'm sorry, I should have realized this and clarified better in the Primitive specification.

The `kick_origin` parameter for the Kick and Chip primitives is where the robot should make the kick from. However, this is *_NOT_* always the location of the current position of the ball. For example, we may tell the robot to kick at a `kick_origin` that's ahead of the ball, so it kicks the ball while moving. It is up to whatever is creating/calling the Primitive to set the `kick_origin` correctly. Most of the time this will be the location of the ball, but sometimes not. 

So the Kick and Chip Primitives don't need the ball information. They just need to make try to shoot at the location given.

(That being said, nice job getting the Ball subscriber set up. You've got everything right. You would just need to add the `Ball` to the constructor of the Visitor)

Ping me on Slack if I can clarify this more

**Comment by MathewMacDougall** [comment]
The issue with the `assignGoalie` function in the `Team` class throwing exceptions has been temporarily fixed in #288. Make sure to update your branch.

**Comment by MathewMacDougall** [comment]
@jessicahe21 any updates?

**Review by HaomiaoZ — COMMENTED** [review]
There is a conflict in `main.cpp`, please fix that before merge. Honestly I'm not sure whether we need to include all individual primitive header files in `main.cpp`. We include them only for test purpose?@MathewMacDougall

File: src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp
Comment (HaomiaoZ) [inline src/thunderbots/software/grsim_communication/visitor/grsim_command_primitive_visitor.cpp]: So the robot will always move from  `(chip_primitive.getChipOrigin()-0.5,chip_primitive.getChipOrigin()-0.5)` to `(chip_primitive.getChipOrigin()+0.8,chip_primitive.getChipOrigin()+0.8)` no matter the direction we are chipping? It seems to me that robot is moving in only one direction no matter which direction we are chipping the ball.

**Comment by MathewMacDougall** [comment]
@HaomiaoZ is correct that we don't need all the primitive headers in `grsim_communication/main.cpp`, but it's not a huge deal so we can clean them up later.

**Comment by MathewMacDougall** [comment]
@jessicahe21 any updates?

**Comment by jessicahe21** [comment]
@HaomiaoZ In grsim the robot does (somehow) do a turn directly behind the destination, it's a little hard to explain, so I made a simple sketch:

cffde06796460b20c22f7dd71d018c63

However, for both `Angle::zero()` and `Angle::half()`, it doesn't seem to work above the red line. It would either go straight past `(0,0)` then start bouncing off the sides of the field like the DVD logo or go back to its original position (the starting position of Yellow Robot 0). 

Although I recall something mentioned about it being normal for one half of the field to not work but I'm not too certain. @MathewMacDougall 

My grsim is receiving stack dumps from the robot team filter as of right now. I did pull from master just this morning so I'm not sure what happened. :-( 

Header files are removed from `main.cpp`.

Thank you both for your patience and giving me extra time.

**Comment by MathewMacDougall** [comment]
I'll try take a look at this tonight and see what's going on.

**Comment by MathewMacDougall** [comment]
@jessicahe21 any luck figuring out the issues?

**Comment by jessicahe21** [comment]
Oops, got a little carried away by the evaluation function. Will update by tonight!

**Comment by jessicahe21** [comment]
After the robot filter fix, the issue remained was that if the robot was located above the red line (see diagram in above comment), it would go back to its "original location" e.g. the starting position of Yellow Robot 0. This was fixed after tweaking the travel distance for the command to move forward to be equal to the command to move backwards!

However I feel like my logic isn't very fool proof and probably flawed. Currently my logic is to send two commands consecutively in order to complete one kick/chip. It goes:

1. Move to a point directly behind the kick/chip origin (I imagine it's like swinging your feet backwards before landing a kick on a soccer ball).

- Set up the point to be 0.5 in x- and y-direction "behind" the kick/chip origin. I think this logic is, in some ways, flawed as Haomiao pointed out. The curves does look smooth in grSim but I think this can be improved.
```
Point destBehind;
destBehind = kick_primitive.getKickOrigin();
destBehind = Point(destBehind.x(), destBehind.y());
destBehind.set(destBehind.x() - 0.5, destBehind.y() - 0.5);
```

- `final_speed_at_destination` is set as `0.0` for both commands. Is this true? I wasn't too sure if this collides with `kick_speed_meters_per_second` at all.

```
motion_controller_command = MotionController::MotionControllerCommand(
            destBehind, kick_primitive.getKickDirection(), 0.0,
            kick_primitive.getKickSpeed(), false, false);
```

2. Move directly forward (0.5 in x- and y-direction) towards the kick/chip origin in order to take the shot.
```
Point destForward;
destForward = kick_primitive.getKickOrigin();
destForward = Point(destForward.x(), destForward.y());
destForward.set(destForward.x() + 0.5, destForward.y() + 0.5);

motion_controller_command = MotionController::MotionControllerCommand(
            destForward, kick_primitive.getKickDirection(), 0.0,
            kick_primitive.getKickSpeed(), false, false);
```

In grSim things look like it's functional but I can't tell if the robot is actually kicking/chipping the ball. 
As well, the robot would arrive at the designation first, THEN rotate to face the correct direction. 

Any suggestions for improvement would be appreciated! Although, since this is a simplified version of the actual intended functionalities (i.e. travel in parabolas, etc), I imagine there might also be limitations.

**Comment by MathewMacDougall** [comment]
Long-winded comment incoming. This should be a bit easier to follow than a bunch of smaller code comments.

You've got the right idea about how to approach this ("winding up" your kick, then taking the kick), but most problems I see are with how you've actually tried to implement it. Don't worry about smoothness for now, lets focus on general correct behavior.

To answer your specific questions:
* `final_speed_at_destination` does not collide with `kick_speed_meters_per_second`. They are entirely separate values. You may want `final_speed_at_destination` to be non-zero, so the robot moves "through" the ball when kicking, but that's a detail you can experiment with later.

It should mostly be a matter of making better use of our geometry types to make the code more general.

Comment 1:
```
destBehind = chip_primitive.getChipOrigin();
destBehind = Point(destBehind.x(), destBehind.y());
```
The second line isn't actually doing anything. The `destBehind` variable is already a `Point` object. Take a look at the `geom/point.h` file for more info on that class.

Edge Case 1:
```
destBehind.set(destBehind.x() - 0.5, destBehind.y() - 0.5);
```
Like Haomiao said, this will cause the robot to always go to the same point relative to the chipOrigin to get "behind" it. What if we are shooting a different direction? The point "behind" the chipOrigin is relative to where we are actually chipping to.

Our `Point` class makes computing this very easy, because we can add Points/Vectors together (like adding lines in 2D space on an x/y plane). If you take a `Point` and add a Vector that points _away_ from where we want to chip, you will get a new point "behind" the origin.

For example `Point c = Point(1, 1) + Vector(-1, 2);` will make `c` equal `Point(0, 3)`. Again, take a look at `geom/point.h`. This should be much easier than using the `set` function. 

That should fix the "swinging your leg backwards" part of the code.

For the "swing your foot to kick" part, you've mostly got the right idea.
```
destForward.set(destForward.x() + 0.5, destForward.y() + 0.5);
```
To match the changes above, you won't need to manually add `0.5` to the values. You should be able to just move back to the `chipOrigin` since you already know that's where you want to be chipping/kicking from. This is where you may want the `final_speed_at_destination` to be some non-zero value so the robot doesn't slow down before kicking.


Finally, your "two-stage" approach is all in the same `if` condition. Sending two commands consecutively like you have know will actually cause the second command to overwrite the first (since it hasn't been able to be sent yet. You haven't left the function!).

```
if ((chip_primitive.getChipOrigin() - robot.position()).len() >= 1)
```
Right now the commands only get done if the robot is at least 1 meter from the chipOrigin (that's what this `if` statement is checking). If the robot is closer than that, nothing actually happens.

You're going to need to break this up to that the robot gets in position to kick (behind the ball), and _then_ moves forwards to kick. I'll leave this up to you.

I hope this clarifies things. Let me know if you have any questions!

File: src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp]: I'm assuming you're trying to rotate the angle with `.quarter()`? This won't work because `quarter()` is a `static` function of `Angle` that's used to _create_ angles, not rotate them.

TIL we don't have a `rotate` function for Angles.

To get the Point you're looking for, you can first create a Point from the kick_direction like you're doing (remove the `quarter()`). Then, the `Point` class has a `perp()` function you can use to get the perpendicular Vector/Point. You'll then need to _add_ this Point to the kick_origin. This will give you a Point on the line perpendicular to the direction of the shot (starting from the kick_origin), which matches how you're using this value below.

Make sure you're checking your code compiles. That will help you tell you're on the right path since it will warn you if functions aren't being used as expected.

File: src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp]: Same comment here as above regarding the use of `half()`.

I'm not entirely sure what this `Point` is supposed to be. Based on how you use it below, you should be able to just use the `kick_origin` instead (to make the line between the dest behind the kick origin, and the kick origin). So maybe this can be removed entirely?

Please correct me if you intended this to be something different.

File: src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp]: I think you just want to use `dest_behind` here rather than `Point(dest_behind_x, dest_behind_y)`. They are not equivalent, because `dest_behind` is relative to the kick_origin.

File: src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp
Comment (MathewMacDougall) [inline src/thunderbots/software/grsim_communication/grsim_command_primitive_visitor.cpp]: I don't think this offset can ever be < 0. If this is intending to check the robot is "in line" with the shot, it would be more robust to check it's less than some small, non-zero value.

The offset is realistically never going to be exactly 0 (and stay exactly 0 as the robot moves).

I also think you meant this to be > 0? If this `if` statement is true, the robot is moving to the `dest_behind`. We don't really need to do this if the robot is already "in line". (Remember that we can assume the preconditions for this Primitive are met).

**Review by MathewMacDougall — CHANGES_REQUESTED** [review]
Looks much better. 

The general structure of your "2 stages" looks good now, you just might want to revise how you're combining the booleans in your `if` statement (think about the cases in your head. Will the robot move behind the ball when expected?)

**Comment by jessicahe21** [comment]
Got a little carried away with working on #351 (near completion, finally!! 😭) But I also understand that getting kick and chip primitives in is a priority as well. 
Will provide a mid-week update, have a tad too much school work and midterms due for Mon/Tues. Thanks!

**Comment by jessicahe21** [comment]
image

@MathewMacDougall The picture here is kind of where I'm going at with the if statements!

If both of the following are not met, move to `dest_behind`:
1. Robot is within 1 robot radius of the red line
2. Robot is further away from the orange line than `dest_behind`

The second one I'm a bit iffy because I think the diagram in #86 depicted "if robot is 1 robot radius further away from the orange line". I tried variations of 
```
offsetToLine(kick_origin, perp_line_to_shot, robot.position()) <= 
ROBOT_MAX_RADIUS_METERS
``` 
but it was not working as intended. Most of the time the robot would be stuck at `dest_behind`.

I'm not too sure why Travis Cl is failing, since locally `catkin_make` and grsim were fine. It said something about `undefined reference to 'offsetToLine(Point, Point, Point)'` although `geom/util.h` was included.

**Comment by MathewMacDougall** [comment]
The reason the robot is get stuck at `dest_behind` is because once the robot tries to move from `dest_behind` towards the ball to take the shot, the code will re-enter the first `if` block because the `offsetToLine(kick_origin, perp_line_to_shot, robot.position()) <= offsetToLine(kick_origin, perp_line_to_shot, dest_behind))` will become true again.

You should be able to simplify this by removing the `offsetToLine(kick_origin, perp_line_to_shot, robot.position()) <= offsetToLine(kick_origin, perp_line_to_shot, dest_behind))` check entirely. That way once the robot is behind the ball, when it moves forwards it will break any conditions of the first `if` statement, so should be able to move forwards successfully.

Travis CI is failing while building the tests. You can do this locally by running `catkin_make tests` or manually choosing the test to build in CLion. It's failing when building the `grsim_communication_test`. This is (probably) because the `grsim_communication_test` is including the `grsim_command_primitive_visitor.*` files that you have now modified to use `geom/util.*`, but the test doesn't include those files. So you should be able to add the `geom/util.cpp` file to the `grsim_communication_test` in the CMakeLists.txt file to fix the build issue.

**Comment by MathewMacDougall** [comment]
@jessicahe21 just ping me when you're ready for re-review

**Comment by MathewMacDougall** [comment]
@jessicahe21 	any updates here? Think we can get this in this week? Pretty sure it just needs some final touch-ups and testing (in grsim).

**Comment by jessicahe21** [comment]
Ah, unit tests for `geom/util` are failing because of the tidbit of changes I made for `offsetToLine`. Working on fixing that right now.

Fix: `abs` -> `fabs`. I never knew `abs` returned `int` type 😣

Ready for re-review 🎉 Thanks!

**Review by MathewMacDougall — APPROVED** [review]
Tried it out locally and it seems to be working as expected :+1: Nice work getting through all this!

**Pull request merged**
